src/storage/src/hummock/sstable/builder.rs

```diff
@@ -126,17 +162,37 @@ impl SstableBuilder {
 
         // TODO: refine me
         value.encode(&mut self.raw_value);
+        let mut extract_key = user_key(full_key);
         if let Some(table_id) = get_table_id(full_key) {
             if self.last_table_id != table_id {
                 self.table_ids.insert(table_id);
                 self.last_table_id = table_id;
             }
         }
+
+        extract_key = self.filter_key_extractor.extract(extract_key);
+
         self.block_builder.add(full_key, self.raw_value.as_ref());
         self.raw_value.clear();
 
-        let user_key = user_key(full_key);
-        self.user_key_hashes.push(farmhash::fingerprint32(user_key));
+        // add bloom_filter check
+        // 1. not empty_key
+        // 2. extract_key key is not duplicate
+        if !extract_key.is_empty()
+            && (extract_key != &self.last_full_key[0..self.last_bloom_filter_key_length])
+        {
+            // avoid duplicate add to bloom filter
+            self.user_key_hashes
+                .push(farmhash::fingerprint32(extract_key));
+            self.last_bloom_filter_key_length = extract_key.len();
+        }
+
+        if self.last_full_key.is_empty() {
```

Comment:
It is not necessary.
And please use `last_full_key` usage in `finish` because we do not need last_full_key in `meta_block`/